sdk/core/azure_core/src/task/mod.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! Asynchronous task execution utilities.
+
+use std::{fmt, future::Future, pin::Pin, sync::Arc};
+
+#[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+mod standard_spawn;
+
+#[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+mod wasm_spawn;
+
+#[cfg(feature = "tokio")]
+mod tokio_spawn;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+pub use standard_spawn::StdSpawner;
+
+#[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+pub use wasm_spawn::WasmSpawner;
+
+#[cfg(feature = "tokio")]
+pub use tokio_spawn::TokioSpawner;
+
+#[cfg(not(target_arch = "wasm32"))]
+pub(crate) type TaskFuture = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+#[cfg(target_arch = "wasm32")]
+pub(crate) type TaskFuture = Pin<Box<dyn Future<Output = ()> + 'static>>;
+
+/// Creates a new [`TaskSpawner`].
+///
+/// This returns a [`TaskSpawner`] that spawns a [`std::thread`] to run the task unless `tokio` was enabled,
+/// in which case an executor is returned that calls [`tokio::task::spawn`].
+pub fn new_task_spawner() -> Arc<dyn TaskSpawner> {
+    #[cfg(feature = "tokio")]
+    {
+        Arc::new(TokioSpawner)
+    }
+    #[cfg(all(not(feature = "tokio"), not(target_arch = "wasm32")))]
+    {
+        Arc::new(StdSpawner)
+    }
+    #[cfg(all(not(feature = "tokio"), target_arch = "wasm32"))]
+    {
+        Arc::new(WasmSpawner)
+    }
+}
+
+#[cfg(feature = "tokio")]
+#[derive(Debug)]
+pub struct SpawnHandle(tokio::task::JoinHandle<()>);
+
+#[cfg(feature = "tokio")]
+impl SpawnHandle {
```

Comment:
The `async_trait` we use everywhere else may help here. It rewrites the signatures into a `BoxedFuture`.

Reply:
Unfortunately, that doesn't work for parameters to traits, only trait functions :(. Because boy, I would love to be able to use it :).